lib/game_runner.rb

```diff
@@ -11,27 +13,40 @@ def initialize(stdout, stdin, grid = Grid.new)
   end
 
   def printer
-    @printer ||= Printer.new(stdout, grid)
+    @printer ||= Printer.new(stdout)
+  end
+
+  def grid_printer
+    @grid_printer ||= GridPrinter.new(stdout, grid)
+  end
+
+  def player_input
+    @player_input ||= PlayerInput.new(stdin, grid, printer)
   end
 
   def run
-    printer.print_welcome_message
-    demand_valid_coordinates
+    print_welcome_message
+    print_grid
+    valid_coordinate
+    mark_grid
+    print_grid
   end
 
   private
 
-  def demand_valid_coordinates
-    while coordinate_invalid?
-      printer.print_coordinates_error
-    end
+  def print_welcome_message
+    printer.print_welcome_message
+  end
+
+  def print_grid
+    grid_printer.print_grid
   end
 
-  def coordinate_invalid?
-    !grid.coordinate_valid?(get_user_input)
+  def valid_coordinate
+    player_input.get_valid_coordinate
   end
 
-  def get_user_input
-    stdin.gets.chomp
+  def mark_grid
+    grid.mark(valid_coordinate)
```

Comment:
Question: Does your flow ask for the user for input twice? 

The run method says `valid_coordinate` and then `mark_grid` which calls `valid_coordinate`. That implies that `valid_coordinate` which seems to ask for user input asks for it twice?

Reply:
Ah interesting! It does not prompt the user for input twice, but I removed it `valid_coordinate` from the run method because it wasn't necessary.